git_services/git_services/init/cloner.py

```diff
@@ -0,0 +1,150 @@
+    """Class for keep track of basic user info used in cloning a repo."""
+
+    username: str
+    full_name: str
+    email: str
+    oauth_token: str
+
+
+class GitCloner:
+    remote_name = "origin"
+    remote_origin_prefix = f"remotes/{remote_name}"
+    autosave_branch_prefix = "renku/autosave"
+
+    def __init__(
+        self,
+        git_url,
+        repo_url,
+        user: User,
+        lfs_auto_fetch=False,
+        repo_directory=".",
+    ):
+        self.git_url = git_url
+        self.repo_url = repo_url
+        repo_directory = Path(repo_directory)
+        if not repo_directory.exists():
+            repo_directory.mkdir(parents=True, exist_ok=True)
+        self.cli = GitCLI(repo_directory)
+        self.user = user
+        self.git_host = urlparse(git_url).netloc
+        self.lfs_auto_fetch = lfs_auto_fetch
+        self._wait_for_server()
+
+    def _wait_for_server(self, timeout_mins=None):
+        start = datetime.now()
+
+        while True:
+            res = requests.get(self.git_url)
+            if res.status_code >= 200 and res.status_code < 400:
+                return
+            if timeout_mins is not None:
+                timeout_tdelta = timedelta(minutes=timeout_mins)
+                if datetime.now() - start > timeout_tdelta:
+                    raise errors.GitServerUnavailableError
+            sleep(5)
+
+    def _initialize_repo(self):
+        self.cli.git_init()
+        self.cli.git_config(f"user.email {self.user.email}")
+        self.cli.git_config(f"user.name {self.user.full_name}")
+        self.cli.git_config("push.default simple")
+
+    def _setup_proxy(self, proxy_url):
+        self.cli.git_config(f"http.proxy {proxy_url}")
+        self.cli.git_config("http.sslVerify false")
+
+    @contextmanager
+    def _temp_plaintext_credentials(self):
+        credential_loc = Path("/tmp/git-credentials")
+        with open(credential_loc, "w") as f:
+            f.write(f"https://oauth2:{self.user.oauth_token}@{self.git_host}")
+        yield self.cli.git_config(f"credential.helper 'store --file={credential_loc}'")
+        # NOTE: Temp credentials MUST be cleaned up on context manager exit
+        credential_loc.unlink()
+        self.cli.git_config("--unset credential.helper")
+
+    def _clone(self, branch):
+        lfs_skip_smudge = "" if self.lfs_auto_fetch else "--skip-smudge"
+        self.cli.git_lfs(f"install {lfs_skip_smudge} --local")
+        self.cli.git_remote(f"add {self.remote_name} {self.repo_url}")
+        self.cli.git_fetch(self.remote_name)
+        res = self.cli.git_checkout(branch)
+        _, err = res.communicate()
+        if res.returncode != 0 or len(err) != 0:
+            if b"no space left on device" in err:
+                # INFO: not enough disk space
+                raise errors.NoDiskSpaceError
+            else:
+                # INFO: the branch simply does not exist, create it and continue
+                self.cli.git_checkout(f"-b {branch}")
```

Comment:
Shouldn't we raise an error if the branch doesn't exist and let the user know?

Or at least not create the branch and just continue on the default branch.

Reply:
@Panaetius this is a good question. I was just directly translating what we do in the shell script. And there is [this](https://github.com/SwissDataScienceCenter/renku-notebooks/blob/master/git-clone/clone.sh#L64) line in the shell script which does the same thing. I will ping Andreas as to why this is the case.

Reply:
I raise an error now - that makes sense. I checked this with Andreas too.